manpages/cpuadoc.man
.TH cpuadoc 1 3.7.0 BL3DEMC
.na
.nh

.SH NAME
cpu.adoc \- Description of computers for parallel processing in eTomo

.SH SYNOPSIS
.nf
.B /usr/local/ImodCalib/cpu.adoc
.fi

.SH DESCRIPTION
The cpu.adoc file contains descriptions of the computers you wish to use when
running parallel processes through eTomo with processchunks(1).  
It is used by eTomo to build the parallel
processing table and to run processchunks.  If you are just using a single
computer with multiple processors, there are two simple alternatives to using
the cpu.adoc file: 1) enable parallel processing and enter the number of
processors in the Settings dialog opened from the eTomo Options menu; or 2)
define the environment variable IMOD_PROCESSORS to be the number of processors
(see imodenv(1)).

The cpu.adoc file
must be placed in the IMOD calibration directory.  The location of 
this directory is defined by the environment variable IMOD_CALIB_DIR, which
has a default value of /usr/local/ImodCalib with a default installation of
IMOD.  You will probably need to create this directory.  Then, prepare a
cpu.adoc file and place it there.  There is an example cpu.adoc file in the
autodoc directory of the IMOD distribution; you should be able to use this as
a starting point.

.SS Computer Sections and Attributes

.TP
.B [Computer = \fIcomputer_name\fB]
Each section describes one computer.  \fIComputer_name\fR will be used by eTomo
and IMOD to contact the computer via
ssh.  As an alternative to listing an actual computer name, the name can simply
be "localhost", in which case programs will be run directly rather through
ssh.  A computer should be listed only once; do not have one section with an
actual name and one section with "localhost".

Computer attributes are name/value pairs found under a computer section, they
have the following syntax: 
.br
.B name = \fIvalue\fR
.br
The order of the attributes
is not important.  All attributes are optional; some are entered in order to
change the default value of an attribute, while others are purely descriptive.
For the descriptive
attributes, you can put anything in them that you want.
However, any attribute you do not use will be omitted from the table, so not
using attributes can save screen space in Etomo.  If all of the users are
familiar with the attributes of the computers or you are planning to use all
of the computers every time you run a parallel process, then you do not
need to use anything but the number attribute.  

.TP
.B number = \fInumber_of_CPUs
The number attribute is necessary if you have a computer with multiple
processors.  If you do not enter this attribute, a single processor is assumed.

.TP
.B gpu\fR[\fB.local\fR]\fB = 1
This attribute means that the GPU (graphics processing unit) of a video card installed in
this computer is available for running compatible IMOD processes.  If the \fBlocal\fR
modifier is used, then the GPU is only available to an eTomo that is being run on this
computer.  Only video cards with CUDA can be used in this way.

.TP
.B gpu.device = \fIGPU#\fR[\fB,\fIGPU#\fR...]
Use this attribute instead of \fBgpu = 1\fR when multiple GPUs are available.
GPUs will be added to the processchunks call in the same order as they are
listed in this attribute.

.TP
.B type = \fICPU_type
This attribute goes into the table under the heading \fBCPU Type\fR and can
contain a description of the computer's processor.

.TP
.B exclude-interface = recon\fR|\fBjoin\fR|\fBpp\fR|\fBpeet
Excludes a computer from the Parallel Processing table depending on the
interface in use.  Interfaces:  Tomogram Reconstruction - \fBrecon\fR, Join -
\fBjoin\fR, Parallel Processing - \fBpp\fR, PEET - \fBpeet\fR.  Only one
interface can be excluded per computer.

.TP
.B users = \fIuser_name\fR[\fB,\fIuser_name\fR...]
A list of the users who have login priviledges on the computer.  If a user is
not on this list, the computer will be excluded from the Parallel Processing
table when the user runs Etomo.  This attribute is optional.  If it is omitted,
the computer will not be excluded based on user name.

.TP
.B speed = \fICPU_speed
This attribute goes into the table under the heading \fBSpeed\fR and can
contain the speed of the computer.  If you want to put units for this
attribute in the column header, see the Global Attributes section below.

.TP
.B memory = \fIamount_of_RAM
This attribute goes into the table under the heading \fBRAM\fR and can contain
the amount of memory the computer has.  If you want to put units for this
attribute in the column header, see the Global Attributes section below.

.TP
.B os = \fIoperating_system
This attribute goes into the table under the heading \fBOS\fR and can contain a
description of the operating system the computer is running.

.TP
.B mountname = \fIname_for_mounting_this_computers_directories_remotely
This attribute specifies the name substituted for \fB%mountname\fR
in a mount rule (see below) when running on this current computer as the
.br
.B mountrule.#.local = \fIlocal_directory_path\fR
.br
.B mountrule.#.local = \fIremote_directory_path\fR
.br
See the section below on Mount Rules.  Mount rules can be entered as
computer attributes in order to specify a rule that applies only for a 
specific machine or to override a global mount rule.

.SS Queue Sections and Attributes

.TP
.B [Queue = \fIqueue_name\fB]
Each section describes one queue in a cluster.  \fIQueue_name\fR will be
displayed in the parallel processing table and sent to
processchunks with the -Q option, but it does not need to match the actual
name of a queue.  \fBComputer\fR section attributes are all valid in
the \fBQueue\fR sections.

.TP
.B command = \fIcommand_to_run_queuechunk\fR
Required.  \fBQueuechunk\fR is a script in IMOD that allows eTomo to work
with specific types of cluster software, but can be modified or replaced to
work with any cluster software.  Se the man page for Queuechunk(1) for the 
specifications that such a script must meet.  If the version in IMOD is
modified or replaced, it should be given a different name, and placed
somewhere on the IMOD user's path other than IMOD/bin.  That name would be
used in place of "queuechunk" in the command defined in this section.

A PBS example where there is only one queue:
.br
.B command = \fRqueuechunk -t pbs

A PBS example where there are multiple queues:
.br
.B command = \fRqueuechunk -t pbs -q \fIqueue_name

This command will be passed in quotes to processchunks.  It will also be used
to get the load from the queue.

.SS Global Attributes

Global attributes are name/value pairs found at the top of the file, before
any computer sections.  They
typically have the following syntax: 
.br
.I nameA.nameB = value
.br
The order of the attributes is not important.

.TP
.B Version = 1.2
The Version attribute is usually required.  It refers to the syntax used 
in .adoc files and should be set to 1.2.

.TP
.B units.speed = \fIunits_for_the_speed_attributes
The value appears in sub-heading row in the Etomo parallel processing table
under the heading \fBSpeed\fR.

.TP
.B units.memory = \fIunits_for_the_memory_attributes
The value appears in sub-heading row in the Etomo parallel processing table
under the heading \fBRAM\fR.

.TP
.B units.load = \fIqueue_load_header\fR[\fB,\fIqueue_load_header\fR...]
Shows how many comma-separated values will be returned by \fBqueuechunk\fR when
it returns a queue's load and provides header(s) for them in the Etomo parallel
processing table.  The default is "Load".

.TP
.B max.tilt = \fImaximum_number_of_CPUs_users_should_use_for_tilt
When this attribute is in use, a reminder not to use too many CPUs will appear
next to the tilt parallel processing checkbox.  This number must be a whole
number.  The purpose of this attribute is to discourage users from using too
many CPUs and from experiencing diminishing returns because
of bottlenecks.  We are currently setting \fBmax.tilt\fR to 12, but your mileage
will vary.

.TP
.B max.volcombine = \fImaximum_number_of_CPUs_users_should_use_for_volcombine
Similarly to \fBmax.tilt\fR, this attribute will cause a recommendation on the
maximum number of CPUs to appear next to the volcombine parallel processing
checkbox.  This step is much more susceptible than tomogram generation 
to diminishing returns from I/O bottlenecks.  We are currently setting 
\fBmax.volcombine\fR to 8.

.TP
.B separate-chunks = \fIvalue
This attribute can be used to force programs to write chunks into separate
files that will be reassembled at the end of processing.  Without this
setting, different instances of the Tilt program will all write to one output
file simultaneously, which has given problems in one Mac installation.  Any
value other than 0 activates the feature.

.TP
.B min.nice = \fIminimum_nice_value
The minimum value of the Nice spinner.  The default is 0.

.TP
.B users-column = \fR[\fB0\fR|\fB1\fR]
When this attribute is present and not set to 0, the Users column will be
included in the Parallel Processing table.

.TP
.B mountrule.#.local = \fIlocal_directory_path\fR|\fIremote_directory_path
See the section below on Mount Rules.

.SS Mount Rules for Local to Remote Path Translations
In order to use parallel processing in IMOD, all computers must be able to 
access the directory where the data and command files are located.  However,
it is not necessary that the directory be referred to by the same name on
the different computers.  When these names differ, you must provide eTomo
with information about how to translate the current working directory path on
the local computer into a path that can be used to access the directory on the
remote computers.  This gets tricky because the true path of a directory, as
revealed by a \fBpwd\fR command, may not be the same as the path that the user
enters to get there.  Thus, in setting up path translations, you need to
change to a typical directory and then use \fBpwd\fR to find out what the
official path to the directory is.  This is the path that eTomo will see on
the local machine, so you need to work out how this needs to be translated
so that it can be accessed on the remote machines.

As a simple example, each Linux machine in the BL3DEMC used to have a directory
named \fB/localscratch\fR which was accessed from any machine as
\fB/scratch/\fIcomputer_name\fR (where \fIcomputer_name\fR is the name of a
machine, without any domain).  The required mount rules were entered as:

.B mountrule.1.local = \fB/localscratch
.br
.B mountrule.1.remote = \fB/scratch/%mountname

Where \fB%mountname\fR is entered exactly as written and will be substituted
for the appropriate mount name.  In our example, the mount name is just the
computer name, but a mount name different from the computer name can be 
entered for an individual computer using the \fBmountname\fR attribute.

For a complicated example, we had a Macintosh running OSX 10.4, and it mounted
our Linux home directories (\fB/home\fR, \fB/home1\fR, \fB/home2\fR) under the
same names.  It mounted the Linux machine scratch directories under
\fB/scratch/\fIcomputer_name\fR.  However, when we were running on the Mac and
\fBcd'd\fR to a user's home directory and entered \fBpwd\fR, we got, e.g.,
\fB/private/var/automount/home1/\fIusername\fR.  When we \fBcd'd\fR to a Linux
scratch directory and entered \fBpwd\fR, we got
\fB/private/var/automount/\fIcomputer_name\fR.

.br
The correct translations can be accomplished with:

.ft B
.nf
mountrule.2.local = /private/var/automount/home
mountrule.2.remote = /home
mountrule.3.local = /private/var/automount
mountrule.3.remote = /scratch
.ft R
.fi


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)

10/9/13: Fixed a problem in eTomo running under Java 1.7 where some titles
where greyed out.(Bug# 1752)

10/1/13: Prevent the "Use Raptor Result..." button in eTomo - Fiducial Model
dialog from being triggered multiple times.(Bug# 1727)

10/1/13: Saving the state of the eTomo - Fiducial Model dialog.(Bug# 1722)

10/1/13: Improved the handling of opening eTomo - Post Processing at the wrong
time.(Bug# 1712)

9/27/13: Added predefined tomodataplots graphs to various places in eTomo.
(Bug# 1707)

9/24/13: IMOD_4-6-30
9/24/13: Prevented tomodataplots from putting out an empty file and Onegenplot
from going into an infinite loop on an empty or inadequate file.


Etomo/src/etomo/storage/CpuAdoc.java
package etomo.storage;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import etomo.BaseManager;
import etomo.storage.autodoc.AutodocFactory;
import etomo.storage.autodoc.ReadOnlyAttribute;
import etomo.storage.autodoc.ReadOnlyAutodoc;
import etomo.storage.autodoc.ReadOnlySection;
import etomo.storage.autodoc.SectionLocation;
import etomo.type.AxisID;
import etomo.type.ConstEtomoNumber;
import etomo.type.EtomoNumber;
import etomo.util.EnvironmentVariable;

/**
 * <p>Description:</p>
 * <p>Represents the cpu.adoc file except the mount rules, which
 * are handled by RemotePath.</p>
 * 
 * <p>Assumptions:</p>
 * <p>The Computer section names in cpu.adoc must be unique.</p>
 * 
 * <p>Copyright: Copyright 2006</p>
 *
 * <p>Organization:
 * Boulder Laboratory for 3-Dimensional Electron Microscopy of Cells (BL3DEMC),
 * University of Colorado</p>
 * 
 * @author $Author$
 * 
 * @version $Revision$
 * 
 * @threadsafe
 * @immutable
 * @singleton
 */
public class CpuAdoc {
  public static final String rcsid = "$Id$";

  public static final String COMPUTER_SECTION_TYPE = "Computer";
  public static final String QUEUE_SECTION_TYPE = "Queue";
  static final String GPU_KEY = "gpu";
  private static final String SPEED_KEY = "speed";
  private static final String MEMORY_KEY = "memory";
  private static final int MIN_NICE_DEFAULT = 0;

  public static final CpuAdoc INSTANCE = new CpuAdoc();

  private final List computerList = new ArrayList();
  private final Map computerMap = new Hashtable();
  private final List queueList = new ArrayList();
  private final Map queueMap = new Hashtable();
  private final EtomoNumber minNice = new EtomoNumber();
  private final EtomoNumber maxTilt = new EtomoNumber();
  private final EtomoNumber maxVolcombine = new EtomoNumber();

  private boolean separateChunks = false;
  private boolean usersColumn = false;
  private String speedUnits = "";
  private String memoryUnits = "";
  private String gpuSpeedUnits = "";// Not in use - has not been added to cpuadoc man page
  private String gpuMemoryUnits = "";// Not in use - has not been added to cpuadoc man
                                     // page
  private String[] loadUnits = new String[0];
  private boolean envVar = false;
  private boolean userConfig = false;
  private boolean loaded = false;
  private boolean exists = false;

  private CpuAdoc() {
    minNice.setDisplayValue(MIN_NICE_DEFAULT);
    minNice.setDefault(MIN_NICE_DEFAULT);
  }

  public boolean isSeparateChunks(BaseManager manager, AxisID axisID,
      String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return separateChunks;
  }

  public boolean isUsersColumn(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return usersColumn;
  }

  public int getMinNice(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return minNice.getInt();
  }

  public String getSpeedUnits(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return speedUnits;
  }

  public String getMemoryUnits(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return memoryUnits;
  }

  public String[] getLoadUnits(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return loadUnits;
  }

  public ConstEtomoNumber getMaxTilt(BaseManager manager, AxisID axisID,
      String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return maxTilt;
  }

  public ConstEtomoNumber getMaxVolcombine(BaseManager manager, AxisID axisID,
      String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return maxVolcombine;
  }

  /**
   * Returns true if there is no computerList entries, or false if any of the computerList
   * entries have a non-local-only GPU.
   * @param manager
   * @param axisID
   * @param propertyUserDir
   * @param ignoredNode if this node is non-null and present in the list, the list is still considered empty.  Parameter may be null
   * @return
   */
  boolean isGpuComputerListEmpty(BaseManager manager, AxisID axisID,
      String propertyUserDir, final Node ignoredNode) {
    load(manager, axisID, propertyUserDir);
    if (computerList.isEmpty()) {
      return true;
    }
    for (int i = 0; i < computerList.size(); i++) {
      Node node = (Node) computerMap.get(computerList.get(i));
      if ((ignoredNode == null || node != ignoredNode) && node.isGpu()) {
        return false;
      }
    }
    return true;
  }

  boolean isComputerListEmpty(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return computerList.isEmpty();
  }

  boolean isQueueListEmpty(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return queueList.isEmpty();
  }

  /**
   * Return computerMap.get(computerList[index]) or null.
   * @param index
   * @param axisID
   * @param propertyUserDir
   * @param managerKey
   * @return
   */
  Node getComputer(BaseManager manager, int index, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return (Node) computerMap.get(computerList.get(index));
  }

  /**
   * Get the Computer section for the current computer.
   * @param axisID
   * @param propertyUserDir
   * @param managerKey
   * @return
   */
  Node getLocalHostComputer(BaseManager manager, AxisID axisID, String propertyUserDir) {
    // Search for "localhost":
    Node localHost = getComputer(manager, Node.LOCAL_HOST_NAME, axisID, propertyUserDir);
    if (localHost != null) {
      return localHost;
    }
    // Search for the computer name:
    String localHostName = Network.getLocalHostName(manager, axisID, propertyUserDir);
    localHost = getComputer(manager, localHostName, axisID, propertyUserDir);
    if (localHost != null) {
      return localHost;
    }
    // Local host not found. Try removing everything from the local host name
    // starting with the first ".".
    int index = localHostName.indexOf('.');
    if (index != -1) {
      localHostName = localHostName.substring(0, index);
      localHost = getComputer(manager, localHostName, axisID, propertyUserDir);
    }
    return localHost;
  }

  Node getQueue(BaseManager manager, int index, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return (Node) queueMap.get(queueList.get(index));
  }

  /**
   * Get queue from queueMap by name.  Return null if not found.
   * @param name
   * @param axisID
   * @param propertyUserDir
   * @param managerKey
   * @return
   */
  Node getQueue(BaseManager manager, String name, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return (Node) queueMap.get(name);
  }

  /**
   * Get computer from computerMap by name.  Return null if not found.
   * @param name
   * @param axisID
   * @param propertyUserDir
   * @param managerKey
   * @return
   */
  Node getComputer(BaseManager manager, String name, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return (Node) computerMap.get(name);
  }

  int getComputerListSize(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return computerList.size();
  }

  int getQueueListSize(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return queueList.size();
  }

  boolean exists(BaseManager manager, AxisID axisID, String propertyUserDir) {
    load(manager, axisID, propertyUserDir);
    return exists;
  }

  private synchronized void load(BaseManager manager, AxisID axisID,
      String propertyUserDir) {
    if (loaded) {
      return;
    }
    loaded = true;
    ReadOnlyAutodoc autodoc = getAutodoc(manager, axisID);
    if (autodoc != null && autodoc.exists()) {
      exists = true;
      separateChunks = loadBooleanAttribute(autodoc, "separate-chunks");
      loadAttribute(minNice, autodoc, "min", "nice");
      usersColumn = loadBooleanAttribute(autodoc, "users-column");
      ReadOnlyAttribute attrib = autodoc.getAttribute("max");
      if (attrib != null) {
        loadAttribute(maxTilt, attrib, "tilt");
        loadAttribute(maxVolcombine, attrib, "volcombine");
      }
      attrib = autodoc.getAttribute("units");
      if (attrib != null) {
        speedUnits = loadStringAttribute(attrib, SPEED_KEY);
        memoryUnits = loadStringAttribute(attrib, MEMORY_KEY);
        loadUnits = loadStringListAttribute(attrib, "load");
        attrib = attrib.getAttribute(GPU_KEY);
        if (attrib != null) {
          gpuSpeedUnits = loadStringAttribute(attrib, SPEED_KEY);
          gpuMemoryUnits = loadStringAttribute(attrib, MEMORY_KEY);
        }
      }
      loadComputers(autodoc);
      loadQueues(autodoc);
    }
  }

  private String loadStringAttribute(ReadOnlyAttribute attrib, String key) {
    if (attrib == null) {
      return "";
    }
    attrib = attrib.getAttribute(key);
    if (attrib == null) {
      return "";
    }
    return attrib.getValue();
  }

  private void loadAttribute(EtomoNumber number, ReadOnlyAttribute attrib, String key) {
    number.reset();
    if (attrib == null) {
      return;
    }
    attrib = attrib.getAttribute(key);
    if (attrib == null) {
      return;
    }
    number.set(attrib.getValue());
  }

  private String[] loadStringListAttribute(ReadOnlyAttribute attrib, String key) {
    if (attrib == null) {
      return new String[0];
    }
    attrib = attrib.getAttribute(key);
    if (attrib == null) {
      return new String[0];
    }
    String list = attrib.getValue();
    if (list == null) {
      return new String[0];
    }
    return list.split("\\s*,\\s*");
  }

  private ReadOnlyAutodoc getAutodoc(BaseManager manager, AxisID axisID) {
    ReadOnlyAutodoc autodoc = null;
    try {
      autodoc = AutodocFactory.getInstance(manager, AutodocFactory.CPU, axisID);
    }
    catch (FileNotFoundException e) {
      e.printStackTrace();
    }
    catch (IOException e) {
      e.printStackTrace();
    }
    catch (LogFile.LockException e) {
      e.printStackTrace();
    }
    if (autodoc == null) {
      System.err.println("Missing $" + EnvironmentVariable.CALIB_DIR
          + "/cpu.adoc file.\n" + "Parallel processing cannot be used.\n"
          + "See $IMOD_DIR/autodoc/cpu.adoc.");
    }
    return autodoc;
  }

  private void loadComputers(ReadOnlyAutodoc autodoc) {
    SectionLocation location = autodoc.getSectionLocation(COMPUTER_SECTION_TYPE);
    if (location == null) {
      return;
    }
    ReadOnlySection section = null;
    while ((section = autodoc.nextSection(location)) != null) {
      Node computer = Node.getComputerInstance();
      computer.load(section);
      String name = section.getName();
      if (computer != null) {
        computerList.add(name);
        computerMap.put(name, computer);
      }
    }
  }

  private void loadQueues(ReadOnlyAutodoc autodoc) {
    SectionLocation location = autodoc.getSectionLocation(QUEUE_SECTION_TYPE);
    if (location == null) {
      return;
    }
    ReadOnlySection section = null;
    while ((section = autodoc.nextSection(location)) != null) {
      Node queue = Node.getQueueInstance();
      queue.load(section);
      String name = section.getName();
      if (queue != null) {
        queueList.add(name);
        queueMap.put(name, queue);
      }
    }
  }

  private boolean loadBooleanAttribute(ReadOnlyAutodoc autodoc, String key) {
    ReadOnlyAttribute attrib = autodoc.getAttribute(key);
    if (attrib != null && (attrib.getValue() == null || !attrib.getValue().equals("0"))) {
      return true;
    }
    return false;
  }

  private void loadAttribute(EtomoNumber number, ReadOnlyAutodoc autodoc, String key1,
      String key2) {
    number.reset();
    ReadOnlyAttribute attrib = autodoc.getAttribute(key1);
    if (attrib == null) {
      return;
    }
    attrib = attrib.getAttribute(key2);
    if (attrib == null) {
      return;
    }
    number.set(attrib.getValue());
  }

  private String loadStringAttribute(ReadOnlyAutodoc autodoc, String key1, String key2) {
    ReadOnlyAttribute attrib = autodoc.getAttribute(key1);
    if (attrib == null) {
      return "";
    }
    attrib = attrib.getAttribute(key2);
    if (attrib == null) {
      return "";
    }
    return attrib.getValue();
  }

  private String[] loadStringListAttribute(ReadOnlyAutodoc autodoc, String key1,
      String key2) {
    ReadOnlyAttribute attrib = autodoc.getAttribute(key1);
    if (attrib == null) {
      return new String[0];
    }
    attrib = attrib.getAttribute(key2);
    if (attrib == null) {
      return new String[0];
    }
    String list = attrib.getValue();
    if (list == null) {
      return new String[0];
    }
    return list.split("\\s*,\\s*");
  }
}
/**
 * <p> $Log$
 * <p> Revision 1.23  2011/06/28 21:09:37  sueh
 * <p> Bug# 1480 Added isGpuComputerListEmpty.
 * <p>
 * <p> Revision 1.22  2011/06/23 14:59:11  sueh
 * <p> Bug# 1495 In getLocalHostComputer added a search for "localhost".
 * <p>
 * <p> Revision 1.21  2011/04/04 16:56:54  sueh
 * <p> bug# 1469 Added/modified GPU_KEY, gpuMemoryUnits, gpuSpeedUnits, MEMORY_KEY, load (refactored),
 * <p> loadAttribute, loadStringAttribute, loadStringListAttribute.  Removed UNITS_KEY.
 * <p>
 * <p> Revision 1.20  2011/02/22 04:32:04  sueh
 * <p> bug# 1437 Reformatting.
 * <p>
 * <p> Revision 1.19  2010/11/04 21:02:25  sueh
 * <p> bug# 1415 Changed MIN_NICE_DEFAULT to 0.
 * <p>
 * <p> Revision 1.18  2010/02/17 04:49:31  sueh
 * <p> bug# 1301 Using the manager instead of the manager key do pop up
 * <p> messages.
 * <p>
 * <p> Revision 1.17  2010/01/11 23:56:47  sueh
 * <p> bug# 1299 Removed responsibility anything other then cpu.adoc from
 * <p> CpuAdoc.  Placed responsibility for information about the network in the
 * <p> Network class.
 * <p>
 * <p> Revision 1.16  2009/03/17 00:44:22  sueh
 * <p> bug# 1186 Pass managerKey to everything that pops up a dialog.
 * <p>
 * <p> Revision 1.15  2009/02/04 23:29:40  sueh
 * <p> bug# 1158 Changed id and exceptions classes in LogFile.
 * <p>
 * <p> Revision 1.14  2008/07/19 00:28:49  sueh
 * <p> bug# 1125 Get rid of isSetByUser and isValid because they where always
 * <p> used together.  Have file, envVar, and userConfig booleans.  IsAvailable
 * <p> is true when one of these booleans is true.  SettingsDialog can also check
 * <p> whether file or envVar is true.  Added setting userConfig and the number
 * <p> of CPUs.
 * <p>
 * <p> Revision 1.13  2008/01/25 22:23:44  sueh
 * <p> bug# 1070 Added setByUser which is true when the user sets the cpu data with
 * <p> cpu.adoc or IMOD_PROCESSORS.
 * <p>
 * <p> Revision 1.12  2007/11/06 19:28:32  sueh
 * <p> bug# 1047 Allow parallel processing even when that is no cpu.adoc and
 * <p> IMOD_PROCESSORS is not set.
 * <p>
 * <p> Revision 1.11  2007/09/27 20:30:38  sueh
 * <p> bug# 1044 Retrieving Queue sections.
 * <p>
 * <p> Revision 1.10  2007/08/29 22:42:32  sueh
 * <p> bug# 1040 Fixing a bug in load(AxisID, BaseManager) where the value max
 * <p> volcombine value was loaded into maxTilt.
 * <p>
 * <p> Revision 1.9  2007/07/17 21:25:47  sueh
 * <p> bug# 1018 Move data reading and storage to CpuAdoc, except mount
 * <p> rules.
 * <p>
 * <p> Revision 1.8  2007/05/22 21:07:57  sueh
 * <p> bug# 999 Added class Computer, to hold section level data.  Added users.
 * <p>
 * <p> Revision 1.7  2007/05/21 22:29:13  sueh
 * <p> bug# 1000 Added excludeInterface and loadExcludeInterfaceMap().
 * <p>
 * <p> Revision 1.6  2007/05/21 18:16:27  sueh
 * <p> bug# 992 Fixed a bug in loadBoolean(); wasn't handle attrib.getValue()
 * <p> returning null.
 * <p>
 * <p> Revision 1.5  2007/05/21 18:10:27  sueh
 * <p> bug# 964 Added usersColumn.
 * <p>
 * <p> Revision 1.4  2007/05/18 23:52:22  sueh
 * <p> bug# 987 Made CpuAdoc thread-safe.  Added minNice.
 * <p>
 * <p> Revision 1.3  2007/03/21 18:10:49  sueh
 * <p> bug# 964 Moved Adoc classes out of the autodoc package because
 * <p> they not part of the autodoc.
 * <p>
 * <p> Revision 1.5  2007/03/15 21:46:20  sueh
 * <p> bug# 964 Added ReadOnlyAttribute, which is used as an interface for Attribute,
 * <p> unless the Attribute needs to be modified.
 * <p>
 * <p> Revision 1.4  2007/03/01 01:19:30  sueh
 * <p> bug# 964 Added LogFile to Autodoc.
 * <p>
 * <p> Revision 1.3  2006/07/21 22:11:49  sueh
 * <p> bug# 901 Getting the calibration directory environment variable name from
 * <p> EnvironmentVariable.
 * <p>
 * <p> Revision 1.2  2006/06/30 17:02:15  sueh
 * <p> Improved warning about missing cpu.adoc.
 * <p>
 * <p> Revision 1.1  2006/06/14 00:33:47  sueh
 * <p> bug# 852 Moved classes to the autodoc package that parse an autodoc or find
 * <p> attributes specific to a type of autdoc.
 * <p>
 * <p> Revision 1.1  2006/06/08 19:04:38  sueh
 * <p> bug# 867 Class to read the cpu autodoc.
 * <p> </p>
 */
